schedutils/coreset.c

```diff
@@ -0,0 +1,282 @@
+ * with this program; if not, write to the Free Software Foundation, Inc.,
+ * 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Copyright (C) 2024 Phil Auld <pauld@redhat.com>
+ */
+
+#include <stdio.h>
+#include <stdlib.h>
+#include <unistd.h>
+#include <getopt.h>
+#include <errno.h>
+#include <sched.h>
+#include <stddef.h>
+#include <sys/prctl.h>
+
+#include "c.h"
+#include "closestream.h"
+#include "optutils.h"
+#include "strutils.h"
+
+// These are in prctl.h on systems new enough. My dev system isn't
+// should check for this in .configure step and not build
+// this?
+
+#ifndef PR_SCHED_CORE
+/* Request the scheduler to share a core */
+# define PR_SCHED_CORE                   62
+#  define PR_SCHED_CORE_GET              0
+#  define PR_SCHED_CORE_CREATE           1 /* create unique core_sched cookie */
+#  define PR_SCHED_CORE_SHARE_TO         2 /* push core_sched cookie to pid */
+#  define PR_SCHED_CORE_SHARE_FROM       3 /* pull core_sched cookie to pid */
+#  define PR_SCHED_CORE_MAX              4
+#endif
+
+/* some prctl.h have the above but not these */
+#ifndef PR_SCHED_CORE_SCOPE_THREAD
+# define PR_SCHED_CORE_SCOPE_THREAD             0  /* PIDTYPE_PID  */
+# define PR_SCHED_CORE_SCOPE_THREAD_GROUP       1  /* PIDTYPE_TGID */
+# define PR_SCHED_CORE_SCOPE_PROCESS_GROUP      2  /* PIDTYPE_PGID */
+#endif
+
+/* basic operation to perform */
+enum cmd_type {
+	CORE_SHOW,    /* this just does PR_SCHED_CORE_GET */
+	CORE_CREATE,  /* PR_SCHED_CORE_CREATE */
+	CORE_PUSH,    /* PR_SCHED_CORE_SHARE_TO */
+	CORE_COPY,    /* PR_SCHED_CORE_SHARE_FROM */
+};
+
+struct coreset {
+	pid_t		pid;		/* task PID (or tid) */
+	unsigned long   cookie;         /* storage for current cookie */
+	int             cmd;            /* what to do: one of cmd_type */
+	int             scope;          /* one of PR_SCHED_CORE_SCOPE_THREAD(0), PR_SCHED_CORE_SCOPE_THREAD_GROUP(1)
+                                         * or PR_SCHED_CORE_SCOPE_PROCESS_GROUP  (2)
+                                         */
+};
+
+static void __attribute__((__noreturn__)) usage(void)
+{
+	FILE *out = stdout;
+	fprintf(out,
+		_("Usage: %s [options] [-p pid] [cmd [args...]]\n\n"),
+		program_invocation_short_name);
+
+	fputs(USAGE_SEPARATOR, out);
+	fputs(_("Show or change the core scheduling cookie for a process or thread.\n"), out);
+	fputs(USAGE_SEPARATOR, out);
+
+	fprintf(out, _(
+		"Options:\n"
+		" -c, --copy              coply the cookie from give pid to this cmd\n"
+		" -n, --new               create new cookie on pid or cmd\n"
+		" -t, --to                copy current task's cookie to existing pid or cmd\n"
+		" Absence of one of the mutually exclusive above options just reports current cookie on given pid (or cmd)\n"
+		" -p, --pid               operate on existing given pid/tid\n"
+		" -s, --scope             0, 1 or 2: apply change to task (0), thread group (1) or process group (2) of given pid/tid\n"
```

Comment:
Wouldn't it be nicer if the user can specify the name of the scope?

Reply:
The real constants for those magic numbers are very long. Seem like a lot of complexity (and user typing) for little gain to me.  They'd need to be validated and then translated back to numbers.

Reply:
I didn't mean the full defines but `task`/`thread-group`/`process-group`

Reply:
That still seems like a lot more typing that 0,1 or 2. But if there is optutils code that can help validate that sort of argument  I'll consider it. I can write code from scratch to do it but seems like it complicates things for little benefit still.  I'd want to accept either the string or the int.

Reply:
The common pattern is just to use a comparison in the getopt loop.

```c
if (strcmp(optarg, "task") == 0
   context = TASK;
else if (strcmp(optarg)...
```